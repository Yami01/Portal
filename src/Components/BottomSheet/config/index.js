/**
 *
 * @format
 * @flow
 *
 */
export const DEFAULT_BOTTOM_PADDING = 16
