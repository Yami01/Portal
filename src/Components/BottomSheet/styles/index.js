/**
 *
 * @format
 * @flow
 *
 */
import { StyleSheet } from 'react-native'
import type { BottomSheetStylesType } from '../types'
import { Colors } from '@/Theme/Variables'

const BottomSheetStyles: BottomSheetStylesType = StyleSheet.create({
  container: {
    backgroundColor: Colors.gray_250,
  },
  handlerWrapper: {
    width: '100%',
    display: 'flex',
    alignItems: 'center',
  },
  handler: {
    marginTop: 8,
    width: 40,
    height: 4,
    borderRadius: 4,
    backgroundColor: Colors.gray_650,
  },
})

export default BottomSheetStyles
