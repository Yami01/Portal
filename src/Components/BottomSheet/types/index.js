/**
 *
 * @format
 * @flow
 *
 */
import type { Node } from 'react'
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet'

export type PropsType = {
  containerStyle?: ViewStyleProp,
  visible: boolean,
  children: Node,
  testID?: string,
  onChange?: number => void,
  backdropTestID?: string,
  contentTestID?: string,
  onDismiss?: () => void,
}

export type BottomSheetStylesType = {
  container: ViewStyleProp,
  handlerWrapper: ViewStyleProp,
  handler: ViewStyleProp,
}
