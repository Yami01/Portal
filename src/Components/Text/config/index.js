/**
 *
 * @format
 * @flow
 *
 */

export const FONT_REGULAR = 'regular'
export const FONT_BOLD = 'bold'
export const FONT_SEMIBOLD = 'semiBold'
export const FONT_LIGHT = 'light'
