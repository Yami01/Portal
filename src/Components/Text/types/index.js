/**
 *
 * @format
 * @flow
 *
 */
import type { Node } from "react"
import type { TextProps } from "react-native/Libraries/Text/TextProps"
import type { TextStyleProp } from "react-native/Libraries/StyleSheet/StyleSheet"

export type PropsType = {
  ...TextProps,
  fontWeight?: 'light' | 'regular' | 'semiBold' | 'bold',
  fontSize?: number,
  color?: string,
  style?: TextStyleProp,
  testID?: string,
  children?: Node,
  allowFontScaling?: boolean,
}

export type TextStylesType = {}
