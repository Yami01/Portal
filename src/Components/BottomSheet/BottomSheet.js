/**
 *
 * BottomSheet
 * @format
 * @flow
 *
 */

import React, {memo, useCallback, useEffect, useMemo, useRef, useState} from 'react';
import type {Node, AbstractComponent} from 'react';
import type {LayoutEvent} from 'react-native/Libraries/Types/CoreEventTypes';
import {View} from 'react-native';
import {BottomSheetModal as LibBottomSheetModal, BottomSheetBackdrop} from '@gorhom/bottom-sheet';

import type {PropsType} from './types';
import {DEFAULT_BOTTOM_PADDING} from './config';
import BottomSheetStyles from './styles';

const BottomSheet: AbstractComponent<PropsType> = memo((props: PropsType): Node => {
  const {
    containerStyle,
    children,
    onChange,
    visible = false,
    testID,
    contentTestID,
    ...restProps
  } = props;
  const bottomSheetModalRef = useRef<any>(null);
  const [modalHeight, setModalHeight] = useState(0);

  useEffect(() => {
    if (visible) {
      bottomSheetModalRef?.current?.present?.();
    } else {
      bottomSheetModalRef?.current?.dismiss?.();
    }
  }, [bottomSheetModalRef, visible]);

  const snapPoints = useMemo((): number[] => [-1, modalHeight], [modalHeight]);

  const onContentLayout = useCallback((event: LayoutEvent) => {
    if (event?.nativeEvent?.layout) {
      const {height} = event?.nativeEvent?.layout;
      setModalHeight(height + DEFAULT_BOTTOM_PADDING);
    }
  }, []);

  const renderHandler = useCallback(
    (): Node => (
      <View style={BottomSheetStyles.handlerWrapper}>
        <View style={BottomSheetStyles.handler} />
      </View>
    ),
    []
  );

  const renderBackdrop = useCallback((backdropProps: any): Node => {
    return <BottomSheetBackdrop opacity={0.3} {...backdropProps} />;
  }, []);

  return (
    <LibBottomSheetModal
      accessibilityViewIsModal
      ref={bottomSheetModalRef}
      index={1}
      snapPoints={snapPoints}
      onChange={onChange}
      handleComponent={renderHandler}
      backdropComponent={renderBackdrop}
      backgroundComponent={null}
      style={containerStyle}
      testID={testID}
      {...restProps}>
      <View testID={contentTestID} onLayout={onContentLayout}>
        {children}
      </View>
    </LibBottomSheetModal>
  );
});

export default BottomSheet;
