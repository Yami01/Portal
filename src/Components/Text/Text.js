/**
 *
 * Text
 * @format
 * @flow
 *
 */

import type { AbstractComponent, Node } from "react"
import React, { memo, useMemo } from "react"
import { Text as LibText } from "react-native"
import type { TextStyleProp } from "react-native/Libraries/StyleSheet/StyleSheet"
import { useTheme } from "react-native-paper"
import { FONT_REGULAR } from "./config"
import type { PropsType } from "./types"

const Text: AbstractComponent<PropsType> = memo((props: PropsType): Node => {
  const {
    fontWeight = FONT_REGULAR,
    fontSize,
    color,
    style,
    children,
    testID,
    ...restProps
  } = props
  const theme = useTheme()

  const textStyle = useMemo((): TextStyleProp => {
    const finalTextStyle: TextStyleProp[] = []

    if (fontWeight) {
      finalTextStyle.push(theme.fonts[fontWeight])
    }
    if (fontSize) {
      finalTextStyle.push({ fontSize })
    }
    if (color) {
      finalTextStyle.push({ color })
    }
    finalTextStyle.push(style)

    return finalTextStyle
  }, [theme, fontSize, fontWeight, color, style])

  return (
    <LibText testID={testID} style={textStyle} {...restProps}>
      {children}
    </LibText>
  )
})

export default Text
