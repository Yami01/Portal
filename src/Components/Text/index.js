/**
 *
 * @flow
 *
 */

import { FONT_BOLD, FONT_LIGHT, FONT_REGULAR, FONT_SEMIBOLD } from "./config"

export { default } from './Text'

export const fontWeightOptions = {
  bold: FONT_BOLD,
  light: FONT_LIGHT,
  semiBold: FONT_SEMIBOLD,
  regular: FONT_REGULAR,
}
