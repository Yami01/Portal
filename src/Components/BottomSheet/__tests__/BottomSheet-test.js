/**
 *
 * BottomSheet Test
 * @format
 * @flow
 *
 */

import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent, act } from '@testing-library/react-native'

import BottomSheet from '../BottomSheet'
import Text from '@/Components/Text'

describe('BottomSheet', () => {
  const props = {
    containerStyle: {},
    onChange: (index: number) => {
      console.log(index)
    },
    testID: 'bottom_sheet_test_id',
    contentTestID: 'bottom_sheet_content_test_id',
  }
  test('BottomSheet with visible props', () => {
    const tree = renderer
      .create(
        <BottomSheet visible {...props}>
          <Text>Testing Bottom Sheet</Text>
        </BottomSheet>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('BottomSheet without visible props', () => {
    const tree = renderer
      .create(
        <BottomSheet visible={false} {...props}>
          <Text>Testing Bottom Sheet</Text>
        </BottomSheet>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('BottomSheet with visible props and trigger onLayout', () => {
    const { getByTestId } = render(
      <BottomSheet visible {...props}>
        <Text>Testing Bottom Sheet</Text>
      </BottomSheet>,
    )
    const bottomSheetView = getByTestId(props.testID)
    const contentView = getByTestId(props.contentTestID)
    act(() => {
      fireEvent(contentView, 'onLayout', {
        nativeEvent: { layout: { height: 300 } },
      })
    })
    expect(bottomSheetView).not.toBeNull()
    expect(contentView).not.toBeNull()
  })
})
