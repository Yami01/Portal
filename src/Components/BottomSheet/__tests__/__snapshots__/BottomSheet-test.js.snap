// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheet BottomSheet with visible props 1`] = `
<View
  accessibilityViewIsModal={true}
  backdropComponent={[Function]}
  backgroundComponent={null}
  handleComponent={[Function]}
  index={1}
  onChange={[Function]}
  snapPoints={
    Array [
      -1,
      0,
    ]
  }
  style={Object {}}
  testID="bottom_sheet_test_id"
>
  <View
    onLayout={[Function]}
    testID="bottom_sheet_content_test_id"
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontWeight": "400",
          },
          undefined,
        ]
      }
    >
      Testing Bottom Sheet
    </Text>
  </View>
</View>
`;

exports[`BottomSheet BottomSheet without visible props 1`] = `
<View
  accessibilityViewIsModal={true}
  backdropComponent={[Function]}
  backgroundComponent={null}
  handleComponent={[Function]}
  index={1}
  onChange={[Function]}
  snapPoints={
    Array [
      -1,
      0,
    ]
  }
  style={Object {}}
  testID="bottom_sheet_test_id"
>
  <View
    onLayout={[Function]}
    testID="bottom_sheet_content_test_id"
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontWeight": "400",
          },
          undefined,
        ]
      }
    >
      Testing Bottom Sheet
    </Text>
  </View>
</View>
`;
